python/tskit/trees.py

```diff
@@ -8684,56 +8684,33 @@ def _genetic_relatedness_vector(
             :rtype: `np.ndarray`
             """
             
-            # maps samples to individuals
-            def sample_individual_sparray(ts: tskit.TreeSequence) -> scipy.sparse.sparray:
-                samples_individual = ts.nodes_individual[ts.samples()]
-                return scipy.sparse.csr_array(
-                        (
-                            np.ones(ts.num_samples),
-                            (np.arange(ts.num_samples), samples_individual)
-                        ),
-                        shape=(ts.num_samples, ts.num_individuals)
-                    )
-            
-            # maps values in idx to num_individuals
-            def individual_idx_sparray(n: int, idx: np.ndarray) -> scipy.sparse.sparray:
-                return scipy.sparse.csr_array(
-                        (
-                            np.ones(idx.size),
-                            (idx, np.arange(idx.size))
-                        ),
-                        shape=(n, idx.size)
-                    )
-                                
             assert cols.size == arr.shape[0], "Dimension mismatch"
-            # centering
+            ij = np.vstack([[n,k] for k, i in enumerate(individuals) for n in self.individual(i).nodes])
+            samples, sample_individuals = ij[:,0], ij[:,1] # sample node index, individual of those nodes
             x = arr - arr.mean(axis=0) if centre else arr # centering within index in rows
-            x = individual_idx_sparray(ts.num_individuals, cols).dot(x)
-            x = sample_individual_sparray(ts).dot(x)
-            x = ts.genetic_relatedness_vector(W=x, windows=windows, mode="branch", centre=False)
-            x = sample_individual_sparray(ts).T.dot(x)
-            x = individual_idx_sparray(ts.num_individuals, rows).T.dot(x)
+            x = self.genetic_relatedness_vector(W=x[sample_individuals], windows=windows, mode="branch", centre=False, nodes=samples)
+            bincount_fn = lambda w: np.bincount(sample_individuals, w)
+            x = np.apply_along_axis(bincount_fn, axis=0, arr=x) # I think it should be axis=1, but axis=0 gives the correct values why?
```

Comment:
The matvec is sometimes GRM * matrix, so `x` is often a matrix than a vector. `np.bincount` only works for 1-dimensional weights, so I used `np.apply_along_axis` and `lambda` to vectorize `np.bincount`.

Reply:
The comment I left after # looks like mostly a convention issue in that function. When `axis=0`, the columns are separately retrieved from the array. When `axis=1`, the rows are retrieved.

Reply:
Agree, that seems confusing but maybe makes sense after all?